Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,3037 @@
+
+    CurveSegmentData::Instance& updDataInst(const State& state) const
+    {
+        return Value<CurveSegmentData::Instance>::updDowncast(
+            getSubsystem().updDiscreteVarUpdateValue(state, m_indexDataInst));
+    }
+
+    const CurveSegmentData::Instance& getPrevDataInst(const State& state) const
+    {
+        return Value<CurveSegmentData::Instance>::downcast(
+            getSubsystem().getDiscreteVariable(state, m_indexDataInst));
+    }
+
+    CurveSegmentData::Instance& updPrevDataInst(State& state) const
+    {
+        return Value<CurveSegmentData::Instance>::updDowncast(
+            getSubsystem().updDiscreteVariable(state, m_indexDataInst));
+    }
+
+    CurveSegmentData::Pos& updDataPos(const State& state) const
+    {
+        return Value<CurveSegmentData::Pos>::updDowncast(
+            getSubsystem().updCacheEntry(state, m_indexDataPos));
+    }
+
+    //------------------------------------------------------------------------------
+    // Data
+    //------------------------------------------------------------------------------
+
+    // Subsystem info.
+    CableSubsystem* m_subsystem = nullptr;
+    // The index of this CurveSegment's cable in the CableSubsystem.
+    CableSpanIndex m_cableIndex = CableSpanIndex::Invalid();
+    // The index of this CurveSegment's obstacle in the cable.
+    ObstacleIndex m_obstacleIndex = ObstacleIndex::Invalid();
+
+    // MobilizedBody that surface is attached to.
+    MobilizedBodyIndex m_body;
+    // Surface to body transform.
+    Transform m_X_BS;
+
+    // Obstacle surface.
+    std::shared_ptr<const ContactGeometry> m_geometry;
+
+    // Topology cache.
+    CacheEntryIndex m_indexDataPos        = CacheEntryIndex::Invalid();
+    DiscreteVariableIndex m_indexDataInst = DiscreteVariableIndex::Invalid();
+
+    // Initial contact point hint used to setup the initial path.
+    Vec3 m_contactPointHint_S{NaN, NaN, NaN};
+
+    //------------------------------------------------------------------------------
+    // Helper class for unit tests.
+    friend CableSubsystemTestHelper;
+};
+
+//==============================================================================
+//                      TODO
+//==============================================================================
+
+namespace
+{
+
+// Given a correction vector computed from minimizing the cost function,
+// compute the maximum allowed stepsize along that correction vector.
+// The allowed step is computed by approximating the surface locally as
+// a circle in each direction, and limiting the radial displacement on that
+// circle.
+void calcMaxAllowedCorrectionStepSize(
+    const CurveSegment& curve,
+    const State& s,
+    const NaturalGeodesicCorrection& c,
+    Real maxAngularDisplacementInDegrees,
+    Real& maxAllowedStepSize)
+{
+    auto UpdateMaxStepSize = [&](Real maxDisplacementEstimate, Real curvature)
+    {
+        const Real maxAngle = maxAngularDisplacementInDegrees / 180. * M_PI;
```

Comment:
I get a compiler error on MSVS saying that `M_PI` is undeclared. Replacing with `SimTK_PI` works.